### PR TITLE
New package: unar-1.10.1

### PR DESCRIPTION
**Pull request by bobertlo**
for request #11339

### CODE REVIEW

**Comment by Duncaen** [comment]
```
build_style=gnu-makefile
make_cmd="make -f Makefile.linux"
build_wrksrc="XADMaster"
make_use_env=yes
```
And no `do_build`.

```
do_install() {
        for x in lsar unar; do
                vbin "$x"
                vman "${wrksrc}/Extra/${x}.1"
                vinstall "${wrksrc}/Extra/${x}.bash_completion" 644 usr/share/bash-completion/completions "$x"
        done
}
```

File: srcpkgs/unar/template [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Template file for 'unar'
+pkgname=unar
+version=1.10.1
+revision=1
+archs="i686 x86_64" # gnustep broken on musl
+wrksrc="kosovan-theunarchiver-1534a5cf6900"
+makedepends="gnustep-make gnustep-base-devel gcc-objc bzip2-devel"
```
Comment (Duncaen) [inline srcpkgs/unar/template]: `gnustep-make` shouldn't be necessary. And I think `gcc-objc` should be in `hostmakedepends`

**Comment by bobertlo** [comment]
@duncaen how do I setup the GNUstep make environment from the "gnu-makefile" build style? will this persist accross build steps?

**Comment by Duncaen** [comment]
Isn't necessary.

File: srcpkgs/unar/template [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+short_desc="Unarchiver for a variety of file formats"
+maintainer="Robert Lowry <bobertlo@gmail.com>"
+license="GPL-2.0-or-later"
+homepage="https://theunarchiver.com/"
+distfiles="https://bitbucket.org/kosovan/theunarchiver/get/${pkgname}-${version}.tar.bz2"
```
Comment (Chocimier) [inline srcpkgs/unar/template]: Please just write `unar` instead of `${pkgname}`.

**Comment by bobertlo** [comment]
@duncaen Ok, that doesn't seem to be necessary. But, it builds code outside of the XADMaster folder, and that code needs to be patches so I think working from the tarball root is still needed.

**Comment by bobertlo** [comment]
If I make it to work in time I'll try to clean up the rest of these comments.

**Comment by Duncaen** [comment]
`build_wrksrc` is made exactly for cases like this, where you have a `wrksrc` and a `build_wrksrc`.

**Comment by DirectorX** [comment]
what about providing `alternatives=` ?

**Comment by Duncaen** [comment]
Alternatives are mainly for things that implement the same thing in a very similar way and are more or less interchangeable.

**Comment by bobertlo** [comment]
@Duncaen Oops, I missed that distinction, thank you for your patience.

I also thought this should just use sed, but then I have been told in the past to use patches instead for similar changes. I'm just going to change it to sed.

I think everything is resolved now.

**Comment by Duncaen** [comment]
I fixed `gnustep-base` for musl, but `unar` is using the `REG_STARTEND` flag for `regexec` which is not supported by musl. This would need a patch that changes the logic or maybe disables the feature.

**Comment by bobertlo** [comment]
@Duncaen thanks for figuring out the dependencies.

I did find this as an example for a C replacement of the non-standard feature:
https://legacy.hylafax.org/archive/1999-03/msg00089.php

I really don't even have time to review that since I don't have that kind of depth in Objective C. I'll throw fixing this for musl on the back burner, but if someone can fix this and wants to maintain it that's cool. Or even with just a patch I could probably maintain it if not :joy: 

Sorry to the requester that this is stuck on musl porting.

**Comment by Johnnynator** [comment]
You can try to use libpcre on musl, it might be less work for patching since you won't need to adjust code logic

File: srcpkgs/unar/template [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# Template file for 'unar'
+pkgname=unar
+version=1.10.1
+revision=1
+archs="i686 x86_64" # gnustep broken on musl
```
Comment (Johnnynator) [inline srcpkgs/unar/template]:
```suggestion
archs="~*-musl" # gnustep broken on musl
```

**Comment by Johnnynator** [comment]
```diff
diff --git a/srcpkgs/unar/patches/musl.patch b/srcpkgs/unar/patches/musl.patch
new file mode 100644
index 0000000000..1fe1cab621
--- /dev/null
+++ b/srcpkgs/unar/patches/musl.patch
@@ -0,0 +1,21 @@
+--- XADMaster/XADRegex.h	2016-05-19 09:18:32.000000000 +0200
++++ -	2019-04-27 12:02:19.093569917 +0200
+@@ -3,7 +3,11 @@
+ #ifdef _WIN32
+ #import "regex.h"
+ #else
++#ifdef __GLIBC__
+ #import <regex.h>
++#else
++#import <pcreposix.h>
++#endif
+ #endif
+ 
+ @interface XADRegex:NSObject
+@@ -86,4 +90,4 @@
+ -(void)replaceEveryPattern:(NSString *)pattern usingSelector:(SEL)selector onObject:(id)object;
+ -(void)replaceEveryPattern:(NSString *)pattern usingSelector:(SEL)selector onObject:(id)object options:(int)options;
+ 
+-@end*/
+\ No newline at end of file
++@end*/
diff --git a/srcpkgs/unar/template b/srcpkgs/unar/template
index 9352e16518..190c7df325 100644
--- a/srcpkgs/unar/template
+++ b/srcpkgs/unar/template
@@ -21,6 +21,11 @@ post_extract() {
 	vsed -i -e '/-D_NATIVE_OBJC_EXCEPTIONS/d' ${wrksrc}/*/Makefile.linux
 }
 
+case $XBPS_TARGET_MACHINE in
+	*-musl) makedepends+=" pcre-devel"
+		LDFLAGS=" -lpcreposix";;
+esac
+
 do_install() {
 
 	for x in unar lsar; do
```

**Comment by bobertlo** [comment]
@Johnnynator that seems like a good idea, cleaner than hacking around the feature.

Unfortunately I tested it on musl and I'm getting `Uncaught exception NSMallocException, reason: Out of memory when creating regex "^(.*)\.(rar|[r-z][0-9]{2})$"` so something is not quite right ;)

**Comment by bobertlo** [comment]
I just reverted this to non-musl only. Maybe I will have more time to work on this.

**Pull request merged**